src/raft.rs

```diff
@@ -888,12 +1002,106 @@ impl<T: Storage> Raft<T> {
     /// according to the progress recorded in r.prs().
     pub fn bcast_append(&mut self) {
         let self_id = self.id;
+        let mut leader_group_id = 0;
+        // Use leader replication if follower replication is disabled or
+        // the broadcast group id of leader is unknown. Broadcast MsgAppend
+        // as normal.
+        let mut use_leader_replication = !self.follower_replication();
+        if !use_leader_replication {
+            leader_group_id = self
+                .prs()
+                .get(self_id)
+                .map_or(0, |pr| pr.broadcast_group_id);
+            use_leader_replication = leader_group_id == 0;
+        }
+        if use_leader_replication {
+            let core = &mut self.r;
+            let msgs = &mut self.msgs;
+            self.prs
+                .iter_mut()
+                .filter(|&(id, _)| *id != self_id)
+                .for_each(|(id, pr)| core.send_append(*id, pr, msgs));
+            return;
+        }
+
+        // If follower replication is enabled, MsgAppends sent to the same broadcast group
+        // will be merge into a MsgGroupBroadcast.
+        //
+        // Messages that needs to be forwarded are stored in hashmap temporarily,
+        // and they are grouped by broadcast_group_id of progress.
+        // Messages in msg_group will be pushed to message queue later.
         let core = &mut self.r;
         let msgs = &mut self.msgs;
-        self.prs
-            .iter_mut()
+        let prs = &mut self.prs.progress;
+        let conf = &self.prs.conf;
+        prs.iter_mut()
             .filter(|&(id, _)| *id != self_id)
-            .for_each(|(id, pr)| core.send_append(*id, pr, msgs));
+            .for_each(|(id, pr)| {
+                let mut tmp_msgs = Vec::default();
+                // Let messages be pushed into tmp_vec firstly.
+                core.send_append(*id, pr, &mut tmp_msgs);
+                for msg in tmp_msgs {
+                    // Filter out messages that need to be forwarded into msg_group.
+                    // Other messages are sent directly.
+                    if pr.broadcast_group_id == leader_group_id
+                        || msg.get_msg_type() != MessageType::MsgAppend
+                        || !pr.is_replicating()
+                    {
+                        msgs.push(msg);
+                    } else {
+                        let peer_id = msg.to;
+                        core.msg_group
+                            .entry(pr.broadcast_group_id)
+                            .or_default()
+                            // The agent must be a voter and active recently.
+                            .push((msg, pr.recent_active && conf.voters().contains(peer_id)));
+                    }
+                }
+            });
+
+        // Merge messages in the same broadcast group and send them.
+        for (_, mut group) in core.msg_group.drain() {
```

Comment:
Can merge be done in L1056?

Reply:
I think it might be impossible because the broadcast group is fixed after all progresses are iterated. The merge should start after the broadcast group is fixed.
If we group progress by `broadcast_group_id` first and then iterate them, it is feasible.